csfieldguide/static/interactives/lzss-compression/js/lzss-compression.js

```diff
@@ -33,15 +39,64 @@ function newLineDiv() {
     return line_div;
 }
 
-// Output the encoded message
+// Write a sentence on the before/after compression text size
+function writeEncodeEfficiency(message, encoded_message) {
+    var unencoded_size = message.length;
+    var encoded_size = 0;
+    var flags = 0;
+
+    for (var i=0; i < encoded_message.length; i++) {
+        if (encoded_message[i] == newlineCharacter || encoded_message[i].length == 1) {
+            encoded_size++;
+        } else {
+            // Assuming a 1.5 byte reference value, we know it's a 4 bit length value
+            // This assumption is too low when looking at long encoded texts
+            encoded_size += 2;
+        }
+        flags++; // Need 1-bit flags for whether or not each item is a character or reference
+    }
+
+    // Round flags to an even number of bytes
+    flags = Math.ceil(flags / 8);
+    //encoded_size += flags; //TODO Decide. It is accurate but may confuse users, particularly since those flags aren't mentioned in the content
+
+    var unencoded_size_message = gettext("Base message size:") + " " + interpolate(ngettext('1 Byte', '%s Bytes', unencoded_size), [unencoded_size]);
+    var encoded_size_message = gettext("Approximate encoded size:") + " " + interpolate(ngettext('1 Byte', '%s Bytes', encoded_size), [encoded_size]);
+
+    document.getElementById('base-size').innerHTML = unencoded_size_message;
+    document.getElementById('encoded-size').innerHTML = encoded_size_message;
+}
+
+// Output the encoded message (text box)
+function writeEncodedMessage(encoded_message) {
+    var compressed_text_div = document.getElementById('message-to-decode');
+    var text_to_write = "";
+    var item;
+
+    for (var i=0; i < encoded_message.length; i++) {
+        item = encoded_message[i];
+
+        if (item == newlineCharacter) {
+            text_to_write += "\n";
+        } else if (item.length == 1) { // just a single character)
+            text_to_write += item;
+        } else { // a reference
+            text_to_write += "(" + item[0] + "," + item[1] + ")";
+        }
+    }
+
+    compressed_text_div.value = text_to_write;
+}
+
+// Output the encoded message (visual box)
 function drawEncodedMessage(encoded_message) {
     var compressed_text_div = document.getElementById('lzss-compression-compressed-text');
 
     // create new div for first line of the message
     var line_div = newLineDiv();
 
     var index = 0;
-    for (var i = 0; i < encoded_message.length; i++) {
+    for (var i=0; i < encoded_message.length; i++) {
         var string = encoded_message[i];
         
         if (string == newlineCharacter) {
```

Comment:
on line 111

`if (string.length == 1)` will trigger a TypeError for newline characters because the string is undefined

Reply:
If I add a bunch of newlines after some text it will not update the text representation of the encoded message

![image](https://user-images.githubusercontent.com/16066822/85970503-cda2b880-ba1e-11ea-8215-d6e70aace021.png)

Reply:
Regarding your first comment, there are two issues:
1) the newline character is `:n` so 2 characters, not undefined
2) the `continue` statement means that part of the code wont execute for newlines anyway (this is code from the original so it does work)

However the problem itself is happening, nice spot

Reply:
Wait line 111 is for the visual representation and it succeeds hmmm